templates/definition/charger/teslalogger.yaml

```diff
@@ -0,0 +1,42 @@
+template: teslalogger
+products:
+  - description:
+      generic: TeslaLogger
+requirements:
+  description:
+    en: Open source Tesla data logger https://github.com/bassmaster187/TeslaLogger, works only with "Teslalogger vehicle"
+    de: Open Source Tesla Datenlogger https://github.com/bassmaster187/TeslaLogger, funktioniert nur in Verbindung mit "TeslaLogger Fahrzeug"
+params:
+  - name: url
+    required: true
+    example: http://192.0.2.2
+  - name: port
+    example: 5000
+    default: 5000
+  - name: id
+    description:
+      de: TeslaLogger CarID
+      en: TeslaLogger CarID
+    default: 1
+render: |
+  type: custom
+  status:
+    source: combined
+    charging:
+      source: http
+      uri: {{ .url }}:{{ .port }}/currentjson/{{ .id }}
+      jq: .charging
+    plugged:
+      source: http
+      uri: {{ .url }}:{{ .port }}/currentjson/{{ .id }}
+      jq: if .plugged_in and .TLGeofenceIsHome then true else false end
+  enabled:
+    source: http
+    uri: {{ .url }}:{{ .port }}/currentjson/{{ .id }}
+    jq: .charging
```

Comment:
Das passt nicht: hier gehts drum, ob Ladefreigabe erteilt ist, also charge_amps > 0?

Reply:
Was genau macht enabled? 
Mit der aktuellen Lösung meckert EVCC ab und zu "out of sync"

Reply:
Gibt an ob Laden freigegeben ist. Das Gegenstück zu `enable`.

Reply:
Hm, da könnte man eine Kombinaion nehmen: Ziel SoC unter Aktuellem SoC. Plugged-in wird eh schon ausgewertet.
Was besseres fällt mir nicht ein

Reply:
Darum gehts nicht. Es geht nur darum ob evcc die Ladung freigegeben hat. Im Zweifel Freigabestrom > 0A?

Reply:
Ich probiere es mal

Reply:
also, wir haben:
charge_current_request - Anforderung (von EVCC als Beispiel)
charger_actual_current - Tatsächlischer Ladestrom (z.B. 99% SoC kann der Strom kleiner sein als request)
charging true/false.

Nun, wenn charging=false, dann sind beide obere Stromwerte 0. wenn charging=true, dann sind die Ströme > 0A.
D.h. was macht nicht wirklich Sinn auf den Strom umzusteigen.

Bei der Tesla-Integration nutzt ihr auch das Gleiche, oder?
https://github.com/evcc-io/evcc/blob/master/vehicle/tesla/controller.go#L71C2-L83C3

Reply:
> Bei der Tesla-Integration nutzt ihr auch das Gleiche, oder?

Da macht das der TWC und wir merken uns einfach den gesetzten Status:

```go
// Enabled implements the api.Charger interface
func (c *Twc3) Enabled() (bool, error) {
	return verifyEnabled(c, c.enabled)
}

// verifyEnabled validates the enabled state against the charger status
func verifyEnabled(c api.Charger, enabled bool) (bool, error) {
	if enabled {
		return true, nil
	}

	status, err := c.Status()

	// always treat charging as enabled
	return status == api.StatusC, err
}
```

Reply:
Ja, da haben wir wieder so ein blödes prinzipbedingtes Problem. Dann lass uns doch erstmal das Fahrzeug fertig machen.

Reply:
Ich wchau mal, ob ich im TeslaLogger merken kann was gefordert wird. Das wäre aber nur für EVCC nützlich... hm

Mit "charging" funktioniert richtig gut, nur ab und zu "out of sync" Warnung...

Reply:
Ich muss ehrlich sagen, den Sinn von Enabled habe ich immer noch nicht verstanden :/
EVCC sendet per Enable: Starte
Und Enabled ust die Bestätigung von dem Charger: Ladevorgang gestartet, richtig?
Und wenn ich den Quellcode Ausschnitt von oben richtig verstehe: wenn Enabled=False aber StatusC (also wird geladen) dann es ist es doch True? 2. Return kann z.B. passieren wenn man am Charger direkt ohne EVCC startet, richtig?
Das ist doch genau das was mein charging=true macht.
Out if sync kommt meistens wenn Tesla selbst anfängt zu laden, z.B. per Timer, oder weil unter SoC Ziel oder manuell via App gestartet.
Oder z.B. EVCC versucht zu laden, Tesla weigert sich aber, weil schon Ziel SoC erreicht oder weil er schläft.

Aber das wäre auch an anderen Chargern mit anderen Autos der Fall.
Oder wird da einfach ein Echo erwartet "Ich habe dich gehört, EVCC, tue aber was ich will"?

Reply:
> richtig?

Nein. Enable heisst "du kannst laden". Ob es da tut oder nicht. Enabled ist der Zeitraum währenddessen der Charger freigeschaltet ist, nicht der währenddessen geladen wird.

Reply:
Ok. Tesla erlaubt das laden nur, wenn im Tesla eingestellter Ziel SoC über dem aktuellen ist. Das werde ich dann als Bedingung nehmen. Denn wenn er Ziel SoC erreicht ist, muss man auch nicht mehr laden. Der Ladevorgang stoppt sofort.
Da TeslaLogger Charger nur mit TeslaLogger Vehicle geht, sollte das passen.
Ich probiere mal aus.

Reply:
Nun, EVCC war auf 90% eingestellt, Tesla 80%, und somit kam bei 80% Ladestopp durch Tesla (enabled=false), EVCC hat aber enabled=true erwartet, also Warnung. Aus meiner Sicht ist es richtig und akzeptabel

Reply:
Ja, das ist in der tat doof. 
Was passiert wenn man an einem "richtigen" Charger oder im Auto an Hardware Stop-Taste klickt?
Da sollte doch auch enabled=false kommen, obwohl evcc true erwartet, oder? Evtl. habe ich die Logik immer noch nicht verstanden 🤔

Oder anders gefragt: Wie würde die Lösung aussehen wenn ich es in TeslaLogger einbauen würde. Von welchen Parametern bzw. Befehlen könnte ich "Enabled" ableiten?

Reply:
> Wie würde die Lösung aussehen wenn ich es in TeslaLogger einbauen würde. 

Das wäre wohl eine Option...

> Evtl. habe ich die Logik immer noch nicht verstanden 🤔

Stell Dir das als schaltbare Steckdose vor. Die ist an oder aus, unabhängig davon ob ein Stecker steckt oder das Gerät lädt. evcc würde die Dose zwar nicht einschalten wenn kein Auto dran hängt, solang eins dran ist allerdings schon (soc mag gefallen sein, Klimaanlage lief etc).

Reply:
Kein Auto daran ist ja Status A
Auto dran, nicht laden ist Status B
Auto dran, laden Status C

Wenn SoC gefallen ist, wird das Auto nachladen, StatusC, egal was EVCC denkt/macht

Nun, beim Status A und B erwarte ich, dass die Dose aus ist, beim C aber an. 
Hier also enabled=false bei charging=false 
Enabled=true bei charging=true


Oder ist die an beim B+C?
Hier ist es also bei enabled=false: plugged_in=false
Beim True: plugged_in=true

Ich sehe z.B. keine Sinn warum bei Status A eine Dose an sein sollte.
Da wäre enabled=true einfach immer.

War es nicht gleiches Problem bei OpenWB?

Reply:
Egal was Du intuitiv erwartest- denk an die Schaltsteckdose. Die schaltet sich auch nicht aus nur weil Du das Ladegerät abziehst...

Reply:
> Dann erklär mir mal bitte, wie Du in Status C kommen willst wenn die Steckdose ausgeschaltet ist???
Status B: enable = false, Steckodse ist aus (Schütze in der Wallbox)
Status C: enable = true, denn zum Laden muss man die Steckdose einschalten (Schütze in der Wallbox)
Bei der Steckdose wissen wir auch nicht, dass etwas angeschlossen ist, bei der Wallbox/ dem TeslaLogger schon ;)

So oder so: ich kommen zu meiner Frage zurück: was muss ich in TeslaLogger programmieren, damit EVCC zufrieden ist? Einfach EVCCs enable "merken" und als "enabled" zurückgeben ist wohl nicht der Sinn der Sache, oder? Das könnte man bestimmt direkt in EVCC mit Script und Variablen lösen, oder?

Ich lade an einem mobilen Charger, ohne jegliche Schnittstelle, so ist die Idee auch entstanden über das Tesla zu regeln.
Nun, wenn ich Kabel anstecke und Tesla keinen Timer hat und SoC und Ziel SoC liegt startet der Ladevorgang sofort. Auch wenn EVCC jetzt enabled=false erwarten würde (z.B. zu wenig Überschuss), läuft der Ladevorgang schon, EVCC ist beleidigt, "disabled, but charging" (oder so ähnlich).

D.h. eigentlich müsste man komplette Wallbox Logik nachbauen: merken was die "externen" wollen: enable=false und sobald man Kabel einsteckt Ladevorgang sofort beenden und warten bis enable=true kommt. Dann würde auch enabled von false auf true erst dann wechseln wenn EVCC Ladevorgang erlaubt.

Nun, im Auto/in der TeslaApp kann ich den Ladevorgang aber trotzdem starten -> enable steht immer noch auf false. Also wird der wieder beendet. Ich will aber laden, jetzt. D.h. da müsste ich zuerst EVCC aufrufen um Ladevorgang zu starten. Ist das auch bei den "richtigen" Wallboxen so?

Aber auch hier, kann zu "out of sync" kommen, wenn das Timing nicht passt und EVCC die Daten in der "falschen" Sekunde abruft. Also eigentlich müsste man auch "gefilterten" isCharing Status vorgaukeln, d.h. isCharing = false, obwohl der Ladevorgang beginnt und TeslaLogger ihn aber gleich beenden wird.

So eine Programmierung wird aber wohl länger dauern... Wir wollt ihr das in der Tesla Integration lösen? Oder (noch) gar nicht?

Reply:
> So oder so: ich kommen zu meiner Frage zurück: was muss ich in TeslaLogger programmieren, damit EVCC zufrieden ist? Einfach EVCCs enable "merken" und als "enabled" zurückgeben ist wohl nicht der Sinn der Sache, oder? Das könnte man bestimmt direkt in EVCC mit Script und Variablen lösen, oder?

Aktuell gibts das nicht weil evcc davon ausgeht, dass sich eine WB ein- und ausschalten lässt. evcc stört sich dran wenn es a) nicht den erwarteten Wert bekommen (Wallbox neu gebootet?) oder b) der Wert nicht zu den anderen passt, z.B. disabled but charging.

> Wir wollt ihr das in der Tesla Integration lösen? Oder (noch) gar nicht?

Was meinst Du? Beim TWC machen wir das so:

```go
// Enable implements the api.Charger interface
func (c *Twc3) Enable(enable bool) error {
	// ignore disabling when vehicle is already disconnected
	// https://github.com/evcc-io/evcc/issues/10213
	status, err := c.Status()
	if err != nil {
		return err
	}
	if status == api.StatusA && !enable {
		c.enabled = false
		return nil
	}

	v, ok := c.lp.GetVehicle().(api.ChargeController)
	if !ok {
		return errors.New("vehicle not capable of start/stop")
	}

	err = v.ChargeEnable(enable)
	if err == nil {
		c.enabled = enable
	}

	return err
}
```

und `Enabled()` macht einen Workaround:

```go
// Enabled implements the api.Charger interface
func (c *Twc3) Enabled() (bool, error) {
	return verifyEnabled(c, c.enabled)
}

// verifyEnabled validates the enabled state against the charger status
func verifyEnabled(c api.Charger, enabled bool) (bool, error) {
	if enabled {
		return true, nil
	}

	status, err := c.Status()

	// always treat charging as enabled
	return status == api.StatusC, err
}
```

Mit anderen Worten: in diesem Fall speichern wir einfach den vorgegebenen Wert und geben ihn gleichzeitig an das Fahrzeug weiter (ein/aus). Für die Abfrage geben wir ihn zurück und validieren ihn nochmal gegen den Status der Wallbox.